README.rst
adpi
----

This package provides a class to control the ADPi family.


debian/changelog
python-adpi (1.0.2+deb11u1) bullseye; urgency=low

  * debian/copyright: Update contact information
  * debian/control: Add Uploaders

 -- Masahiro Honda <honda@mechatrax.com>  Mon, 11 Dec 2023 18:11:58 +0900

python-adpi (1.0.2) bullseye; urgency=medium

  * debian
    - Deleted python-adpi.install
  * debian/compat: Deleted
  * debian/control
    - Deleted Build-Depends python(>=2.7), python-smbus, pythonj-spidev, python-setuptools
    - Deleted Package python-adpi
    - Change Build-Depends to debhelper-compat from debhelper
    - Bump Standards-Version to 4.5.1
  * debian/rules
    - Deleted options to build python-adpi
  * setup.py
    - Fix version string

 -- Yuta Kiyonaga <kiyonaga@mechatrax.com>  Tue, 20 Jul 2021 08:16:43 +0000

python-adpi (1.0.1) buster; urgency=medium

  * setup.py: Update install_requires
  * Update debhelper compatibility level to 12
  * Bump Standards-Version to 4.3.0

 -- Masahiro Honda <honda@mechatrax.com>  Mon, 08 Jul 2019 16:25:10 +0900

python-adpi (1.0.0) unstable; urgency=medium

  * Initial release

 -- Masahiro Honda <honda@mechatrax.com>  Thu, 05 Jul 2018 10:22:24 +0900


adpi/ADC/AD779x.py
class AD7794():
    mode = {
        'continuous':   0x00,
        'single':       0x20,
        'idle':         0x40, 
        'powerdown':    0x60,
        'cal_int_zero': 0x80,
        'cal_int_full': 0xA0,
        'cal_sys_zero': 0xC0,
        'cal_sys_full': 0xE0
    }
    gain = {
        '1':   0,
        '2':   1,
        '4':   2,
        '8':   3,
        '16':  4,
        '32':  5,
        '64':  6,
        '128': 7,
    }
    channel = {
        '1':    0x00,
        '2':    0x01,
        '3':    0x02,
        '4':    0x03,
        'temp': 0x06,
    }
    rate = {
        '470': 0x01,
        '242': 0x02,
        '123': 0x03,
        '62':  0x04,
        '50':  0x05,
        '39':  0x06,
        '33':  0x07,
        '19':  0x08,
        '17':  0x09,
        '16':  0x0A,
        '12':  0x0B,
        '10':  0x0C,
        '8':   0x0D,
        '6':   0x0E,
        '4':   0x0F,
    }
    register = {
       'communications': {'address': 0x00, 'byte': 1},
       'status':         {'address': 0x00, 'byte': 1},
       'mode':           {'address': 0x08, 'byte': 2},
       'configuration':  {'address': 0x10, 'byte': 2},
       'data':           {'address': 0x18, 'byte': 3},
       'id':             {'address': 0x20, 'byte': 1},
       'io':             {'address': 0x28, 'byte': 1},
       'offset':         {'address': 0x30, 'byte': 3},
       'fullscale':      {'address': 0x38, 'byte': 3},
    }
    
    def __init__(self, spi):
        self.spi = spi

    def reset(self):
        cmd = [0xFF, 0xFF, 0xFF, 0xFF]
        self.spi.xfer2(cmd)
    
    def read(self, cmd):
        return self.spi.xfer2(cmd)[1:]

    def write(self, cmd):
        self.spi.xfer2(cmd)

    def read_bytes(self, reg):
        cmd = [reg['address'] | 0x40] + [0x00] * reg['byte']
        return self.read(cmd)

    def write_bytes(self, reg, dat):
        cmd = [reg['address']] + dat
        self.write(cmd)

    def read_data(self, reg):
        cmd = [reg['address'] | 0x40] + [0x00] * reg['byte']
        dat = self.read(cmd)
        val = 0
        for i in range(reg['byte']):
            val <<= 8
            val += dat[i]
        return val
        
    def write_data(self, reg, val):
        cmd = [reg['address']] + [0x00] * reg['byte']
        for i in range(reg['byte'], 0, -1):
            cmd[i] = val & 0xFF
            val >>= 8
        self.write(cmd)



adpi/adpi.py
from time import sleep
from .ADC.AD779x import AD7794
from .EEPROM.AT24 import AT24C256
from .GPIO.MCP230xx import MCP23008

class ADPiPro():
    def __init__(self, spi, i2c, eeprom_addr=0x57, gpio_addr=0x27):
        self.adc = AD7794(spi)
        self.eeprom = AT24C256(i2c, eeprom_addr)
        self.gpio = MCP23008(i2c, gpio_addr)
        self.channels = 4

    def reset(self):
        self.adc.reset()
        sleep(0.0005)

    def read_status(self):
        return self.adc.read_data(self.adc.register['status'])

    def read_mode(self):
        return self.adc.read_bytes(self.adc.register['mode'])

    def write_mode(self, mode, rate):
        self.adc.write_bytes(self.adc.register['mode'], [mode, rate])
    
    def read_configuration(self):
        return self.adc.read_bytes(self.adc.register['configuration'])
    
    def write_configuration(self, gain, ch):
        self.adc.write_bytes(self.adc.register['configuration'], [gain, ch])

    def read_data(self):
        return self.adc.read_data(self.adc.register['data'])

    def read_id(self):
        return self.adc.read_data(self.adc.register['id'])

    def read_offset(self):
        return self.adc.read_data(self.adc.register['offset'])

    def write_offset(self, val):
        self.adc.write_data(self.adc.register['offset'], val)

    def read_fullscale(self):
        return self.adc.read_data(self.adc.register['fullscale'])

    def write_fullscale(self, val):
        self.adc.write_data(self.adc.register['fullscale'], val)

    def load_bias(self, idx):
        dat = self.eeprom.read(32 * idx)
        val = []
        for i in range(4):
            val.append(0)
            for j in range(4):
                val[i] += dat[4 * i + j] << (8 * j)
        return val
    
    def save_bias(self, idx, val):
        dat = []
        for v in val:
            for i in range(4):
                dat.append((v >> (8 * i)) & 0xFF)
        self.eeprom.write(32 * idx, dat)
        sleep(0.005)
    
    def load_scale(self, idx):
        dat = self.eeprom.read(32 * idx + 16)
        val = []
        for i in range(4):
            val.append(0)
            for j in range(4):
                val[i] += dat[4 * i + j] << (8 * j)
        return val
    
    def save_scale(self, idx, val):
        dat = []
        for v in val:
            for i in range(4):
                dat.append((v >> (8 * i)) & 0xFF)
        self.eeprom.write(32 * idx + 16, dat)
        sleep(0.005)
   
    def get_output(self, idx):
        olat = self.gpio.read(self.gpio.register['olat'])
        return (olat >> idx) & 0x01
	
    def set_output(self, idx, val):
        olat = self.gpio.read(self.gpio.register['olat'])
        if val:
            olat |= 0x01 << idx
        else:
            olat ^= 0x01 << idx
        self.gpio.write(self.gpio.register['iodir'], 0xF0)
        self.gpio.write(self.gpio.register['olat'], olat)

